crates/matrix-sdk/src/client/mod.rs

```diff
@@ -1411,30 +1411,51 @@ impl Client {
         }
     }
 
+    /// Whether a room recently joined should be marked as a DM, if we were
+    /// invited to it beforehand.
+    async fn must_mark_joined_room_as_dm(&self, room_id: &RoomId) -> bool {
+        if let Some(room) = self.get_room(room_id) {
+            room.state() == RoomState::Invited
+                && room.is_direct().await.unwrap_or_else(|e| {
+                    warn!(%room_id, "is_direct() failed: {e}");
+                    false
+                })
+        } else {
+            false
+        }
+    }
+
     /// Join a room by `RoomId`.
     ///
-    /// Returns a `join_room_by_id::Response` consisting of the
-    /// joined rooms `RoomId`.
+    /// Returns the `Room` in the joined state.
     ///
     /// # Arguments
     ///
     /// * `room_id` - The `RoomId` of the room to be joined.
     pub async fn join_room_by_id(&self, room_id: &RoomId) -> Result<Room> {
         let request = join_room_by_id::v3::Request::new(room_id.to_owned());
         let response = self.send(request).await?;
+
+        let mark_as_dm = self.must_mark_joined_room_as_dm(&response.room_id).await;
         let base_room = self.base_client().room_joined(&response.room_id).await?;
-        Ok(Room::new(self.clone(), base_room))
+        let room = Room::new(self.clone(), base_room);
+        if mark_as_dm {
+            room.set_is_direct(true).await?;
+        }
+
+        Ok(room)
```

Comment:
Can't we put this code snippet into a function? Seems like we have the same thing in two places now, unless I'm missing something.

Reply:
You're right, I've put all of this inside `must_mark_joined_room_as_dm` and renamed it `finish_join_room` :+1: